n2v/methods/oucarter.py

```diff
@@ -0,0 +1,351 @@
+            l_phi_z = np.array(points_func.basis_values()["PHI_Z"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_xx = np.array(points_func.basis_values()["PHI_XX"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_yy = np.array(points_func.basis_values()["PHI_YY"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_zz = np.array(points_func.basis_values()["PHI_ZZ"])[:l_npoints, :l_lpos.shape[0]]
+
+            lD = D[(l_lpos[:, None], l_lpos)]
+            # lC = C[l_lpos, :]
+
+            rho = contract('pm,mn,pn->p', l_phi, lD, l_phi)
+            rho_inv = 1/rho
+
+            # Calculate the second term 0.25*\nabla^2\rho
+            laplace_rho_temp = contract('ab,pa,pb->p', lD, l_phi, l_phi_xx + l_phi_yy + l_phi_zz)
+            # laplace_rho_temp += contract('pm, mn, pn->p', l_phi_x,lD, l_phi_x)
+            # laplace_rho_temp += contract('pm, mn, pn->p', l_phi_y,lD, l_phi_y)
+            # laplace_rho_temp += contract('pm, mn, pn->p', l_phi_z,lD, l_phi_z)
+            laplace_rho_temp += np.sum((l_phi_x @ lD) * l_phi_x, axis=1)
+            laplace_rho_temp += np.sum((l_phi_y @ lD) * l_phi_y, axis=1)
+            laplace_rho_temp += np.sum((l_phi_z @ lD) * l_phi_z, axis=1)
+
+            laplace_rho_temp *= 0.25 * 2
+
+            # Calculate the third term |nabla rho|^2 / 8
+            tauW_temp = contract('pm, mn, pn->p', l_phi, lD, l_phi_x) ** 2
+            tauW_temp += contract('pm, mn, pn->p', l_phi, lD, l_phi_y) ** 2
+            tauW_temp += contract('pm, mn, pn->p', l_phi, lD, l_phi_z) ** 2
+            tauW_temp *= rho_inv * 0.125 * 4
+
+            tauLmP_rho[iw: iw + l_npoints] = (-laplace_rho_temp + tauW_temp) * rho_inv
+            iw += l_npoints
+        assert iw == tauLmP_rho.shape[0], "Somehow the whole space is not fully integrated."
+
+        return tauLmP_rho
+
+    def _get_optimized_external_potential(self, grid_info, average_alpha_beta=False):
+        """
+        (22) in [1].
+        """
+
+        Nalpha = self.nalpha
+        Nbeta = self.nbeta
+
+        # SVWN calculation
+        wfn_LDA = psi4.energy("SVWN/" + self.basis_str, molecule=self.mol, return_wfn=True)[1]
+        Da_LDA = wfn_LDA.Da().np
+        Db_LDA = wfn_LDA.Db().np
+        Ca_LDA = wfn_LDA.Ca().np
+        Cb_LDA = wfn_LDA.Cb().np
+        epsilon_a_LDA = wfn_LDA.epsilon_a().np
+        epsilon_b_LDA = wfn_LDA.epsilon_b().np
+        self.Vpot = wfn_LDA.V_potential()
+
+        vxc_LDA_DFT = self.on_grid_vxc(Da=Da_LDA, Db=Db_LDA, vpot=self.Vpot)
+        vxc_LDA = self.on_grid_vxc(Da=Da_LDA, Db=Db_LDA, grid=grid_info)
+        if self.ref != 1:
+            assert vxc_LDA.shape[-1] == 2
+            vxc_LDA_beta = vxc_LDA[:,1]
+            vxc_LDA = vxc_LDA[:, 0]
+            vxc_LDA_DFT_beta = vxc_LDA_DFT[:, 1]
+            vxc_LDA_DFT = vxc_LDA_DFT[:, 0]
+
+        e_bar_DFT = self._average_local_orbital_energy(Da_LDA, Ca_LDA[:,:Nalpha], epsilon_a_LDA[:Nalpha])
```

Comment:
Can we add a comment here stating that the  _average_local_orbital_energy is taken from the mRKS?

Reply:
Will do.